### PR TITLE
Feature/merge birch

### PR DESCRIPTION
**Pull request by dhedey**
Merges birch from scrypto into the node. Also takes the opportunity to:
* Properly type check all our raw payloads and hashes in Java and Rust
* Avoid re-hashing transactions everywhere
* Get rid of `REv2TestTransactions.java` and replace it with a more configurable, easier to use `TransactionBuilder` and `Manifest` (creator)
* Fix genesis, revert to 5 minute epochs, which were somehow accidentally changed to 100-round epochs in ash

### CODE REVIEW

**Review by jakrawcz-rdx — APPROVED** [review]
Hereby I 🟢 the rust changes to the state manager + DB + mempool (I only have minor comments there).

The changes are massive - some are barely related to the scrypto lib update (more like drive-by refactors) but they are in a good direction. They also uncover more refactoring opportunities (e.g. a separated `DuplicateIntentHashDetector` responsibility) that I have 99% ready on my own branch.

I will most likely not even attempt to merge my changes on top of these (and instead I will re-apply the ideas on the new code). It should be even easier than on the original code (e.g. I see you renamed the accu-hash CF to `TxnIdentifiersByStateVersion` already and this aligns nicely with my longer-term plan).

To sum up: this looks cool, yet **still someone else should review** the java changes (ofc after you finish propagating them through jni) AND core api.

File: core-rust/state-manager/src/transaction/ledger_transaction.rs [hunk 1]
```diff
@@ -1,43 +1,360 @@
...
+}
+
+impl HasLegacyLedgerPayloadHash for ValidatedLedgerTransaction {
+    fn legacy_ledger_payload_hash(&self) -> LegacyLedgerPayloadHash {
+        self.legacy_ledger_payload_hash
+    }
+}
+
+// A hash of the whole payload, for use by the accumulator
+// TODO: Remove
```
[the remark this answers is not in the document]
Comment (dhedey) [reply core-rust/state-manager/src/transaction/ledger_transaction.rs:327]: 100% agreed with you! Will fix :)

File: core-rust/state-manager/src/transaction/executable_logic.rs [hunk 2]
```diff
@@ -137,7 +137,7 @@ where
                 "Transaction execution took {}ms, above warning threshold of {}ms ({})",
                 elapsed.as_millis(),
                 self.threshold.as_millis(),
-                self.logged_description
+                (self.logged_description)()
```
[the remark this answers is not in the document]
Comment (dhedey) [reply core-rust/state-manager/src/transaction/executable_logic.rs:140]: Hah yes - this was a drive-by-refactor which got out of hand :).

It's not really an issue - just felt weird to have a string created for the purpose of logging if it isn't needed in the main path.

Can revert :)

File: core-rust/core-api-server/src/core_api/handlers/lts/transaction_construction.rs [hunk 3]
```diff
@@ -14,23 +13,22 @@ pub(crate) async fn handle_lts_transaction_construction(
...
+        CONSENSUS_MANAGER.as_node_id(),
+        &ConsensusManagerField::CurrentTime.into(),
     )?;
 
     Ok(models::LtsTransactionConstructionResponse {
-        current_epoch: to_api_epoch(&mapping_context, epoch_manager_substate.epoch)?,
+        current_epoch: to_api_epoch(&mapping_context, consensus_manager_substate.epoch)?,
         ledger_clock: Box::new(to_api_instant_from_safe_timestamp(
-            clock_substate.current_time_rounded_to_minutes_ms,
+            timestamp_substate.epoch_milli,
```
[the remark this answers is not in the document]
Comment (dhedey) [reply core-rust/core-api-server/src/core_api/handlers/lts/transaction_construction.rs:31]: Definitely intentional - this was one of the purposes of capturing the accurate timestamp! 👍

File: core-rust/state-manager/src/mempool/mempool_manager.rs [hunk 4]
```diff
@@ -206,58 +212,86 @@ impl MempoolManager {
...
         source: MempoolAddSource,
-        unvalidated_transaction: NotarizedTransaction,
-    ) -> Result<(), MempoolAddError> {
-        // Quick check to avoid transaction validation if it couldn't be added to the mempool anyway
-        let mut write_mempool = self.mempool.write();
-        write_mempool.check_add_would_be_possible(&unvalidated_transaction.user_payload_hash())?;
-        drop(write_mempool);
+        raw_transaction: RawNotarizedTransaction,
+        force_recalculate: bool,
+    ) -> Result<Arc<MempoolTransaction>, MempoolAddError> {
```
[the remark this answers is not in the document]
Comment (dhedey) [reply core-rust/state-manager/src/mempool/mempool_manager.rs:235]: Arcs should be much cheaper to clone that a full transactions (think about it like a thread-safe pointer to the content stored on the heap) - which should speed up all the methods in the Mempool Manager which clone the whole mempool! - or need access to a particular tranasction.

We could also consider improving these methods to iterate inside the mempool to reduce cloning the Arcs at all if not needed.

File: core-rust/state-manager/src/transaction/executable_logic.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jakrawcz-rdx) [reply core-rust/state-manager/src/transaction/executable_logic.rs:140]: I'd probably stay with a `String` just for sake of simplicity. If it hurts us, then we should probably do what java loggers do (to avoid eagerly rendering a string): pass some `impl Display` there. Maybe an enum like `TimeWarningSubject`?.
Although of course such instance would also need to be created.
And TBH the `Fn` is also a "lightweight struct" (closure) that has to be created and passed, so there is always a cost.
(and we just hope it's more lightweight than rendering a string, but in practice I am sure the difference is not measurable, given how much other computation happens around - that's why I'd go for `String`'s simplicity)

File: core/src/test-core/java/com/radixdlt/rev2/Manifest.java [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+        String.format(
+            """
+            %s
+            CALL_METHOD Address("%s") "withdraw" Address("%s") Decimal("%s");
+            CALL_METHOD Address("%s") "try_deposit_batch_or_abort" Expression("ENTIRE_WORKTOP");
+            """,
+            params.lockFeeLine(account),
+            params.encode(account),
+            params.encode(XRD),
+            ScryptoConstants.FREE_AMOUNT_FROM_FAUCET.subtract(Decimal.of(50)),
```
Comment (jakrawcz-rdx) [inline core/src/test-core/java/com/radixdlt/rev2/Manifest.java:221]:
```suggestion
            ScryptoConstants.FREE_AMOUNT_FROM_FAUCET.subtract(Decimal.of(1000)),
```

The last failing test (i.e. `later_rejected_transaction_should_not_linger_in_mempool()`) will pass if you adjust this line (I assume `50` was too low to account for lock fees etc.)

BTW the test improvements in this PR are unbelievable 👍

**Review by jakrawcz-rdx — APPROVED** [review]

**Review by LukasGasior1 — APPROVED** [review]

File: core-rust-bridge/src/main/java/com/radixdlt/statecomputer/commit/CommittableTransaction.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+package com.radixdlt.statecomputer.commit;
+
+import com.radixdlt.lang.Option;
+import com.radixdlt.sbor.codec.CodecMap;
+import com.radixdlt.sbor.codec.StructCodec;
+import com.radixdlt.transactions.*;
+import com.radixdlt.utils.UInt32;
+
+public record CommittableTransaction(
+    Option<UInt32> index,
```
[the remark this answers is not in the document]
Comment (dhedey) [reply core-rust-bridge/src/main/java/com/radixdlt/statecomputer/commit/CommittableTransaction.java:74]: I actually had it like that to start with, but found it slightly uglier to use in practice. Maybe it's best though happy to change if you'd rather :)

File: core-rust-bridge/src/main/java/com/radixdlt/statecomputer/commit/ActiveValidatorInfo.java [hunk 7]
```diff
@@ -65,14 +65,13 @@
...
 
 import com.radixdlt.crypto.ECDSASecp256k1PublicKey;
-import com.radixdlt.lang.Option;
 import com.radixdlt.rev2.ComponentAddress;
 import com.radixdlt.rev2.Decimal;
 import com.radixdlt.sbor.codec.CodecMap;
 import com.radixdlt.sbor.codec.StructCodec;
 
 public record ActiveValidatorInfo(
-    Option<ComponentAddress> address, ECDSASecp256k1PublicKey key, Decimal stake) {
```
[the remark this answers is not in the document]
Comment (dhedey) [reply core-rust-bridge/src/main/java/com/radixdlt/statecomputer/commit/ActiveValidatorInfo.java:75]: Do you know why it was an option? I just couldn't see any reason why?

File: core-rust-bridge/src/main/java/com/radixdlt/statecomputer/commit/CommittableTransaction.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (LukasGasior1) [reply core-rust-bridge/src/main/java/com/radixdlt/statecomputer/commit/CommittableTransaction.java:74]: happy to keep it as is if that's what turned out less ugly :smile:
